File: Frontend/src/services/userService.js
import api from "@/api/axios";

export const userService = {

    // =========================
    // Admin
    // =========================

    async GetUsers(page = 1, pageSize = 10) {
        const response = await api.get(
            `/users?page=${page}&pageSize=${pageSize}`
        );

        return response.data;
    },

    async GetUserById(id) {
        const response = await api.get(
            `/users/${id}`
        );

        return response.data;
    },

    async CreateAdmin(userData) {
        const response = await api.post(
            "/users/admins",
            userData
        );

        return response.data;
    },

    async CreateTeacher(userData) {
        const response = await api.post(
            "/users/teachers",
            userData
        );

        return response.data;
    },
    async getTeachers(){
        const response = await api.get("users/teachers");
        return response.data;
    },
    async activeUser(userId) {
        const response = await api.put(`users/${userId}/active`);
        return response;
    },
    async getStudents(page = 1, size = 10) {
        const response = await api.get("users/students", {
            params: {
                        page: page,
                        pageSize: size 
                    }
                    });
        return response.data;
},
    // =========================
    // Student
    // =========================


    async getMyInformation(){
        const response = await api.get(
            "/users/me"
        );
        return response.data;
    },
    async UpdateMyProfile(userData) {
        const response = await api.put(
            "/users/me",
            userData
        );

        return response.data;
    },

    async deleteMyAccount() {
        const response = await api.delete(
            "/users/me"
        );

        return response.data;
    },
    async deleteUser(userId){
      const response = await api.delete(`users/${userId}/delete`);
        return response;
    }

};

File: Backend/Business/Services/UserService.cs
using Business.Interfaces;
using Business.Interfaces.Repository;
using Microsoft.AspNetCore.Identity;
using Models.DTOs.Pagination;
using Models.DTOs.Student;
using Models.DTOs.Teacher;
using Models.DTOs.User;
using Models.Entities;


namespace Business.Services
{
    public class UserService : IUserService
    {
        private IUnitOfWork _unitOfWork;
        private IPasswordService _passwordService;
        public UserService(IUnitOfWork unitOfWork,IPasswordService passwordService)
        {
            _unitOfWork = unitOfWork;
            _passwordService = passwordService;
        }
        public async Task AssignRole(int userId, int roleId)
        {
            await _unitOfWork.Users.AssignRole(userId, roleId);
            
        }

        public async Task CreateUser(UserCreateDto dto)
        {
            // we will make the hashing for password 
            // there here for creating teacher or admin
            dto.Password =_passwordService.HashPassword(dto.Password);
           
            await _unitOfWork.Users.AddUserAsync(dto);

            await _unitOfWork.SaveChangesAsync();
           
        }
        public  async Task DeleteUser(int userId)
        {
            await _unitOfWork.Users.DeleteUser(userId);
        }
        public async Task ActiveUser(int userId)
        {
            await _unitOfWork.Users.Active(userId);
        }

        public  Task<IEnumerable<UserDto>> GetAllUsers(int pageNumber,int pageSize)
        {
            return  _unitOfWork.Users.GetUsersAsync(pageNumber,
             pageSize);
        }

        public  Task<UserDto> GetUserByEmail(string email)
        {
            return  _unitOfWork.Users.GetUserByEmail(email);
        }

        public  Task<UserDto> GetUserById(int userId)
        {
            return  _unitOfWork.Users.GetUserById(userId);
        }

        public async Task UpdateUser(UserUpdateDto dto)
        {
            await  _unitOfWork.Users.UpdateUser(dto);
            await _unitOfWork.SaveChangesAsync();
        }
        public async Task<IEnumerable<TeacherDto>> GetTeachersAsync()
        {
           return await _unitOfWork.Users.GetTeachers();
        }
        public async Task<PaginatedResult<StudentDto>> GetStudentsAsync(int page, int size)
        {
            return await _unitOfWork.Users.GetStudents(page,size);
        }
    }
   
    public class PasswordService : IPasswordService
    {
        private readonly PasswordHasher<User> _hasher = new();

        public string HashPassword(string password)
        {
            return _hasher.HashPassword(null!, password);
        }

        public bool VerifyPassword(string providedPassword, string hashedPassword)
        {
            var result = _hasher.VerifyHashedPassword(
                null!,
                hashedPassword,
                providedPassword);

            return result == PasswordVerificationResult.Success
                || result == PasswordVerificationResult.SuccessRehashNeeded;
        }
    }
}
